### PR TITLE
ENH: `integrate.cubature`: support for infinite limits

### PR DESCRIPTION
**Pull request by ollybritton**
#### Reference issue
- #20252
- Builds on:
	- #21330
	- #21473
- Originally part of #21632, which is now being split into multiple PRs so it is easier to review.

#### What does this implement/fix?
This adds support to `cubature` for approximating integrals with one or more sets of infinite limits:

$$
\int^\infty_{-\infty} \int^\infty_{-\infty} e^{-x^2-y^2} \text dx \text dy
$$

```python
cubature(gaussian, [-np.inf, -np.inf], [np.inf, np.inf])
```

This is achieved by applying the same variable transformations used in `quad_vec` to each variable separately. Integrating over intervals of the form $[a, \infty)$, $(-\infty, b]$ is also supported.

This PR also adds supports for specifying problematic points in the integrand's domain through a `points` argument:

$$
\int^1_{-1} \frac{\sin x}{x} \text dx
$$

```python
cubature(sinc, [-1], [1], points=[[0]])
```

These additions maintain support for array API, with the exception of JAX, as the implementation of the variable transformations use in-place index assignment.

### Acknowledgements
Thanks to @izaid and @ev-br, who have provided lots of help implementing this functionality, and thanks to @lucascolley for help with array API support.

### CODE REVIEW

File: scipy/integrate/_cubature.py [hunk 1]
```diff
@@ -114,6 +114,12 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
         map-like callable, such as :meth:`python:multiprocessing.pool.Pool.map` for
         evaluating the population in parallel. This evaluation is carried out as
         ``workers(func, iterable)``.
+    points : list of array_like, optional
+        List of points by which to split up the initial region of integration. If the
+        rule being used does not evaluate `f` on the boundary of a region (which is the
+        case for all Genz-Malik and Gauss-Kronrod rules) then `f` will not be evaluated
+        at these points. This should be a list of array-likes where each element has
+        length ``ndim``. Default is empty. See Examples.
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:122]: The example is useful but I think an extra sentence here too wouldn't hurt explaining why `points` not being evaluated here is useful

File: scipy/integrate/_cubature.py [hunk 2]
```diff
@@ -172,6 +178,24 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
...
+    If :math:`a_i` and :math:`b_i` range over :math:`x \in [a_i, \infty)`, the i-th
+    integration variable will use the transformation :math:`x = a_i + \frac{1-t}{t}` and
+    :math:`t \in (0, 1)`.
+
+    If :math:`a_i` and :math:`b_i` range over :math:`x \in (-\infty, b_i]`, the i-th
+    integration variable will use the transformation :math:`x = b_i - \frac{1+t}{t}` and
+    :math:`t \in (0, 1)`.
+
+    In all three of these cases, the Jacobian of the transformation is
+    :math:`J(t) = t^{-2}`.
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:197]: I think we could probably skip this detail, it is perhaps slightly confusing as there are n variables so the Jacobian is n by n, although I get the point it is making.

File: scipy/integrate/_cubature.py [hunk 3]
```diff
@@ -258,13 +282,46 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
...
+    >>> res.estimate
+     1.8921661
     """
 
     # It is also possible to use a custom rule, but this is not yet part of the public
     # API. An example of this can be found in the class scipy.integrate._rules.Rule.
 
     xp = array_namespace(a, b)
     max_subdivisions = float("inf") if max_subdivisions is None else max_subdivisions
+    points = [] if points is None else [xp.asarray(p, dtype=xp.float64) for p in points]
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:324]: Is there a reason for not using the default floating type? or maybe something using `xp.result_type`

File: scipy/integrate/_cubature.py [hunk 4]
```diff
@@ -298,10 +355,45 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
...
+
+        points = [f.inv(point) for point in points]
+        points.extend(f.points)
+
+    # If any problematic points are specified, divide the initial region so that these
+    # points lie on the edge of a subregion.
+    #
+    # This means ``f`` won't be evaluated there if the rule being used has no evaluation
+    # points on the boundary.
+    if points == []:
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:380]: prehaps more idiomatic?
```suggestion
    if len(points) == 0:
```

File: scipy/integrate/_cubature.py [hunk 5]
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
     err_sub = rule.estimate_error(f, a_k_sub, b_k_sub, args)
 
     return a_k_sub, b_k_sub, est_sub, err_sub
+
+
+def _is_strictly_in_region(point, a, b):
+    xp = array_namespace(point, a, b)
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:475]: Can we add `xp` as an argument to skip the `array_namespace` call?

File: scipy/integrate/_cubature.py [hunk 6]
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
...
+
+def _split_region_at_points(a, b, points):
+    """
+    Given the integration limits `a` and `b` describing a rectangular region and a list
+    of `points`, find the list of ``[(a_1, b_1), ..., (a_l, b_l)]`` which breaks up the
+    initial region into smaller subregion such that no `points` lie strictly inside
+    any of the subregions.
+    """
+
+    xp = array_namespace(a, b)
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:491]: same here

File: scipy/integrate/_cubature.py [hunk 7]
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
...
+
+            else:
+                new_subregions.append((a_k, b_k))
+
+        regions = new_subregions
+
+    return regions
+
+
+class _VariableTransform:
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:525]: Are there other transforms planned?

File: scipy/integrate/_cubature.py [hunk 8]
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
...
+    If :math:`a_i` and :math:`b_i` range over :math:`x \in (-\infty, b_i]`, the i-th
+    integration variable will use the transformation :math:`x = b_i - \frac{1+t}{t}` and
+    :math:`t \in (0, 1)`.
+
+    In all three of these cases, the Jacobian of the transformation is
+    :math:`J(t) = t^{-2}`.
+    """
+
+    def __init__(self, f, a, b):
+        self._xp = array_namespace(a, b)
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:602]: Consider passing `xp` as an argument

File: scipy/integrate/_cubature.py [hunk 9]
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
...
+        a = xp_copy(self._orig_a)
+        b = xp_copy(self._orig_b)
+
+        for index in self._double_inf_pos:
+            a[index] = -1
+            b[index] = 1
+
+        for index in self._semi_inf_pos:
+            a[index] = 0
+            b[index] = 1
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:647]:
```suggestion
        a[self._double_inf_pos] = -1
        b[self._double_inf_pos] = 1

        a[self._semi_inf_pos] = 0
        b[self._semi_inf_pos] = 1
```

File: scipy/integrate/_cubature.py [hunk 10]
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
...
+        t = xp_copy(x)
+
+        for i in self._negate_pos:
+            t[..., i] *= -1
+
+        for i in self._double_inf_pos:
+            t[..., i] = 1/(x[..., i] + self._xp.sign(x[..., i]))
+
+        for i in self._semi_inf_pos:
+            t[..., i] = 1/(x[..., i] - self._orig_a[i] + 1)
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:670]: Again these are arrays right so we can skip the loop and use a list of indexes

File: scipy/integrate/_cubature.py [hunk 11]
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
...
+
+        for i in self._double_inf_pos:
+            # For (-oo, oo) -> (-1, 1), use the transformation x = (1-|t|)/t.
+            x[..., i] = (1 - self._xp.abs(t[..., i])) / t[..., i]
+            jacobian *= 1/(t[..., i] ** 2)
+
+        for i in self._semi_inf_pos:
+            # For (start, oo) -> (0, 1), use the transformation x = start + (1-t)/t.
+            x[..., i] = self._orig_a[i] + (1 - t[..., i]) / t[..., i]
+            jacobian *= 1/(t[..., i] ** 2)
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:689]: same here re indexing

File: scipy/integrate/tests/test_cubature.py [hunk 12]
```diff
@@ -388,6 +475,26 @@ def test_zero_width_limits(self, xp):
...
+            basic_1d_integrand,
+            a,
+            b,
+            args=(n, xp),
+        )
+
+        xp_assert_close(
+            res.estimate,
+            -basic_1d_integrand_exact(n, xp),
+            rtol=1e-1,
```
Comment (j-bowhay) [inline scipy/integrate/tests/test_cubature.py:494]: This is a very loose tol, any reason why?

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Thanks @ollybritton a first pass over things

File: scipy/integrate/_cubature.py (hunk 7 above)
Comment (ollybritton) [reply scipy/integrate/_cubature.py:525]: Yes, integrals with functions as limits are also calculated using a transform. But if you think two transforms aren't enough to justify the base class, then I'm happy to remove this.

File: scipy/integrate/_cubature.py (hunk 2 above)
Comment (ollybritton) [reply scipy/integrate/_cubature.py:197]: Maybe I could rephrase this to "Jacobian determinant"? I think this would make more sense, especially since the details of the transformations above are for just one variable at a time.

Comment (j-bowhay) [reply scipy/integrate/_cubature.py:197]: Sound good

File: scipy/integrate/_cubature.py (hunk 3 above)
Comment (ollybritton) [reply scipy/integrate/_cubature.py:324]: Thanks for the review. I'm a bit confused here, would the "default floating type" be something like `dtype=xp.floating` or `dtype=float`? I'm applying `xp.asarray` to each element in the list so that `points` can be specified as `array_like`. If using `dtype=xp.float64` is a problem here, is it also a problem on the lines below this one?

```
# Convert a and b to arrays
a = xp.asarray(a, dtype=xp.float64)
b = xp.asarray(b, dtype=xp.float64)
```

Comment (j-bowhay) [reply scipy/integrate/_cubature.py:324]: Sorry this was a slightly rushed comment. What I meant was this forces everything to be `xp.float64` whereas in some libraries (pytorch) the default floating dtype is `xp.float32`. If we did something like `a = xp.asarray(a, dtype=xp.results_type(a, b, 1.0))` then we don't promote things unnecessarily. cc @lucascolley @mdhaber who can probably explain more coherently than me

Comment (mdhaber) [reply scipy/integrate/_cubature.py:324]: I think `_array_api.xp_broadcast_promote` with `force_floating=True` may be useful here.
https://github.com/scipy/scipy/blob/ef157bf12f15cbf47f04f2d38576499be5a7703b/scipy/_lib/_array_api.py#L587

**Review by fancidev — COMMENTED** [review]
Good stuff! Just some cosmetic editorial comments.

File: scipy/integrate/_cubature.py (hunk 3 above)
Comment (j-bowhay) [reply scipy/integrate/_cubature.py:324]: It would also be good to add a test for different `dtype`s of `a` and `b` to check the `dtype` of the result is promoted correctly

File: scipy/integrate/_cubature.py [hunk 13]
```diff
@@ -172,6 +178,24 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
...
+    If :math:`a_i` and :math:`b_i` range over :math:`x \in (-\infty, \infty)`, the i-th
+    integration variable will use the transformation :math:`x = \frac{1-|t|}{t}` and
+    :math:`t \in (-1, 1)`.
+
+    If :math:`a_i` and :math:`b_i` range over :math:`x \in [a_i, \infty)`, the i-th
+    integration variable will use the transformation :math:`x = a_i + \frac{1-t}{t}` and
+    :math:`t \in (0, 1)`.
+
+    If :math:`a_i` and :math:`b_i` range over :math:`x \in (-\infty, b_i]`, the i-th
+    integration variable will use the transformation :math:`x = b_i - \frac{1+t}{t}` and
```
[the remark this answers is not in the document]
Comment (ollybritton) [reply scipy/integrate/_cubature.py:193]: I think $b_i - (1+t)/t$ is correct here, since the $(-\infty, b_i]$ case is handled by flipping the limits, negating $x$ and then applying the transformation for $[a_i, \infty)$.

[the remark this answers is not in the document]
Comment (fancidev) [reply scipy/integrate/_cubature.py:193]: But if $0 < t < 1$ and $x = b_i - (1+t)/t$, then $x$ will range from $-\infty$ to $b_i-2$, and there is a “gap” $[b_i-2,b_i]$, right?

File: scipy/integrate/_cubature.py (hunk 3 above)
Comment (ollybritton) [reply scipy/integrate/_cubature.py:324]: Just writing some tests for this and found the following case: Rules like `gk21` return the nodes and weights as arrays using `xp.asarray([0.996, ..., -0.996])` and so are using the default floating dtype in that array namespace.

Since the default floating dtype in Numpy is `float64`, this means that if a user does something like:

```
cubature(
        f,
        np.asarray([0], dtype=np.float32),
        np.asarray([1], dtype=np.float32),
)
```

then the result will actually have dtype `float64` rather than `float32` because calculations are done with the nodes and weights, which themselves will have dtype `float64`. If they were using `torch` instead, then the result would have dtype `float32`. Is this what should happen?

If the nodes and weights should also be returned with dtype `float32` in the NumPy case, then I think it would be necessary to add a new parameter like `result_dtype` to the constructor of each `FixedRule`, since each `Rule` can be initialised independently without `a` or `b`.

File: scipy/integrate/_cubature.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ollybritton) [reply scipy/integrate/_cubature.py:193]: Ah yes, thanks!

File: scipy/integrate/_cubature.py (hunk 3 above)
Comment (mdhaber) [reply scipy/integrate/_cubature.py:324]: > Is this what should happen?

I think the whole calculation should respect the common dtype of the user input. 

> Rules like gk21 return the nodes and weights as arrays using `xp.asarray([0.996, ..., -0.996])` and so are using the default floating dtype in that array namespace.

It sounds like this would mean that even if the user were to provide `float64` input, the _actual_ precision of the output is no better than the default dtype of the array library. Is that right?

I'd suggest storing the nodes and weights with `float64` and downcasting if necessary before use (this is what I do, anyway) or adding a `dtype` parameter.

File: scipy/integrate/_cubature.py (hunk 10 above)
Comment (ollybritton) [reply scipy/integrate/_cubature.py]: I've tried this but am getting errors using `array_api_strict`. Switching to boolean indexing also doesn't work since `t[..., self._semi_inf_pos]` raises errors about not being able to use boolean indexing and `...` at the same time.

File: scipy/integrate/_cubature.py (hunk 2 above)
Comment (ollybritton) [reply scipy/integrate/_cubature.py]: I've now removed this since the Jacobian determinant for the last transformation is $-t^{-2}$, not $t^{-2}$.

**Comment by ollybritton** [comment]
I've just pushed some changes hopefully addressing most of what is above:

- Fix the details of the transformations being used in the docstring
- Pass `xp` to `_split_subregion`, `_is_strictly_in_region` and `_apply_fixed_rule` rather than repeatedly calling `array_namespace`
- Respect the dtype of the inputs when returning results
- Tighten tolerances on some tests

Also, I realised that `_negate_pos` in `_InfiniteLimitsTransform` code was redundant, so I've removed it. Also from above:

- Indexing arrays with a list rather than looping over indices: I'm getting errors using `array_api_strict` for this, and using boolean indexing also isn't working in expressions like `x[..., self._semi_inf_pos]` where `_semi_inf_pos` is a boolean array.
- Nodes and weights are now float64 by default rather than the default floating dtype of the array library.

File: scipy/integrate/_cubature.py (hunk 7 above)
Comment (j-bowhay) [reply scipy/integrate/_cubature.py:524]: If there's stuff to come then I think this is fine

**Comment by lucascolley** [comment]
> using boolean indexing also isn't working in expressions like x[..., self._semi_inf_pos] where _semi_inf_pos is a boolean array.

https://data-apis.org/array-api/draft/API_specification/indexing.html#boolean-array-indexing seems to suggest that, for example, you can replace:

```python
In [13]: x.shape
Out[13]: (2, 2)
...
In [10]: x[..., np.asarray([False, True])]
Out[10]: 
array([[2],
       [4]])
```

with

```python
In [20]: np.expand_dims(x[np.asarray([[False, True], [False, True]])], axis=1)
Out[20]: 
array([[2],
       [4]])
```

I'm not sure how this generalises, but maybe you can do something like that.

**Comment by ollybritton** [comment]
@lucascolley I've just run a very informal comparison on my machine, but it seems like using a for loop is comparable in performance to using boolean arrays in NumPy:

---

- x shape: `(10, 10, 10)`
- ix: `[2, 4, 6]`

|Method|Time|
|------|----|
|`[..., ix]`|8.32 μs ± 90.5 ns|
|`[..., ix_bool]`|9.95 μs ± 51.7 ns|
|for loop: 8.05 μs|89.7 ns|

---

- Similar to what will actually occur with a 5D integrand
- x shape: `(21**5, 5)`
- ix: `[1, 2]`

|Method|Time|
|------|----|
|`[..., ix]`|40.9 ms ± 1.87 ms|
|`[..., ix_bool]`|38.9 ms ± 1.36 ms|
|for loop|13.3 ms ± 203 μs|

---

The notebook I used to check:

```python
data = np.random.rand(21**5, 5)
ix = [1, 2]
ix_bool = [False, True, True, False, False]
```

```python
%%timeit
data[..., ix] *= 2
data[..., ix] /= 2
```

```python
%%timeit
data[..., ix_bool] *= 2
data[..., ix_bool] /= 2
```

```python
%%timeit
for i in ix:
    data[..., i] *= 2
    data[..., i] /= 2
```

Is it worth trying the other method or do you think a for loop over the indicies is acceptable?

File: scipy/integrate/_cubature.py [hunk 14]
```diff
@@ -374,3 +471,211 @@ def _process_subregion(data):
...
+        self._xp = xp
+
+        self._f = f
+        self._orig_a = a
+        self._orig_b = b
+
+        self._semi_inf_pos = []
+        self._double_inf_pos = []
+
+        ndim = xp_size(a)
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:606]: Currently unused
```suggestion
```

Comment (j-bowhay) [reply scipy/integrate/_cubature.py:606]: Oops sorry I confused myself as i was making local edits

File: scipy/integrate/_cubature.py [hunk 15]
```diff
@@ -298,10 +357,45 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
 
             rule = ProductNestedFixed([base_rule] * ndim)
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:358]: Sorry not directly this PR but it would be great to add test coverage to line 356

File: scipy/integrate/_cubature.py [hunk 16]
```diff
@@ -374,3 +471,233 @@ def _process_subregion(data):
...
+
+    regions = [(a, b)]
+
+    for point in points:
+        if xp.any(xp.isinf(point)):
+            # If a point is specified at infinity, ignore.
+            #
+            # This case occurs when points are given by the user to avoid, but after
+            # applying a transformation, they are removed.
+            continue
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:499]: Again not blocking but would it be possible to add a test which hits this line?

File: scipy/integrate/_cubature.py [hunk 17]
```diff
@@ -374,3 +471,233 @@ def _process_subregion(data):
...
+        a[self._semi_inf_pos] = 0
+        b[self._semi_inf_pos] = 1
+
+        return a, b
+
+    @property
+    def points(self):
+        # If there are infinite limits, then the origin becomes a problematic point
+        # due to a division by zero there.
+        if self._num_inf != 0:
```
Comment (j-bowhay) [inline scipy/integrate/_cubature.py:642]: According to the test coverage this condition is always true

**Review by j-bowhay — COMMENTED** [review]
Some minor points but otherwise in good shape

File: scipy/integrate/_cubature.py (hunk 17 above)
Comment (j-bowhay) [reply scipy/integrate/_cubature.py:642]: I think because of the way it is used by `cubature`

File: scipy/integrate/tests/test_cubature.py [hunk 18]
```diff
@@ -872,6 +1260,45 @@ def test_genz_malik_1d_raises_error(self, xp):
...
+        """
+
+        points = [xp.asarray(p, dtype=xp.float64) for p in points]
+
+        f_transformed = _InfiniteLimitsTransform(
+            # Bind `points` and `xp` argument in f
+            lambda x: f_with_problematic_points(x, points, xp),
+            xp.asarray(a, dtype=xp.float64),
+            xp.asarray(b, dtype=xp.float64),
+            xp=xp,
```
Comment (lucascolley) [inline scipy/integrate/tests/test_cubature.py:1292]: @ollybritton here's the culprit for the array API failure. Either `xp` needs to come from `array_namespace` here, or `_InfiniteLimitsTransform` needs to apply `array_namespace` to the input `xp` during setup.

Comment (ollybritton) [reply scipy/integrate/tests/test_cubature.py:1292]: Thanks!

File: scipy/integrate/_cubature.py (hunk 16 above)
Comment (ollybritton) [reply scipy/integrate/_cubature.py:512]: Sorry, should have bumped this earlier. There's now a test which hits this in the latest commit.

File: scipy/integrate/_cubature.py (hunk 15 above)
Comment (ollybritton) [reply scipy/integrate/_cubature.py:358]: Great, I'll make a note of this.

**Comment by j-bowhay** [comment]
@ollybritton, just to let you from tomorrow onwards I'll be away travelling for the next four months without much computer access. I might be able to have a look on the mobile app but hopefully someone else will be able to pickup the review for your final pr for arbitrary regions.

**Comment by ev-br** [comment]
Where does this PR stand w.r.t. review comments?

**Comment by j-bowhay** [comment]
> Where does this PR stand w.r.t. review comments?

I think it is just https://github.com/scipy/scipy/pull/21657#discussion_r1790529005 that needs resolving

File: scipy/integrate/_cubature.py (hunk 17 above)
Comment (ev-br) [reply scipy/integrate/_cubature.py:655]: FWIW, I'd suggest keeping the code as is then, and add a comment. The less footguns for future development, the better.

**Review by j-bowhay — APPROVED** [review]
LGTM provided CI is green!

**Comment by ev-br** [comment]
Congrats @ollybritton ! Do we have a tracking issue for possible follow-ups to the `cubature` work? If not, could you please open one.

**Comment by lucascolley** [comment]
Yeah, I'd suggest closing gh-20252 with a summary of what has been implemented, and opening a new issue with further ideas.
